src/snowflake/snowpark/_internal/udf_utils.py

```diff
@@ -565,6 +571,28 @@ def resolve_imports_and_packages(
             max_batch_size,
             source_code_display=source_code_display,
         )
+
+        if pickled_by_reference:
+            try:
+                source_code = code_generation.generate_source_code(func, False)
+            except Exception as e:
```

Comment:
I think we should raise an exception here to indicate where the failure comes from. The original error message is not very informative. I think a behavior change might be worth it here?

Reply:
I meant we still regard `generate_source_code` as an experimental feature, and we don't use it to generate real source code currently, because there might be some corner cases that we don't realize yet. So I don't think we should roll out this feature just because a bug fix needs it, and if it fails, changing back to the original behavior is better? 

@sfc-gh-aling @sfc-gh-yixie wdyt?

Reply:
If we don't generate source code, will the UDF code fail eventually? This seems to be the case from what I read, so I feel raising an exception is okay because the error message is more informative.

My worry is: given `generate_source_code` is experimental, will there be cases where we generate wrong source code? If not then I think this seems fine as is.

Reply:
is it possible we return the same `ModuleNotFoundError` but with the new message to keep the behavior the same? We could say something like: "module not found, tried to extract source code into a temp module, but errors encountered during module generation thus the module is unable to be generated"..

@sfc-gh-sfan as far as i'm concerned, so far the source code of the function itself shall be good, as it's simply calling `inspect.getsource`. The error-proning part is the references -- classes, global/nonlocal vars, other modules, but I haven't yet seen wrong code being generated, it just fails in case of error.

Reply:
I could raise a `ModuleNotFoundError` instead, but the underlying behavior will still be different: previously the exception is raised by the UDF server trying to register the UDF, now the exception is thrown by Snowpark Python before even submitting the generated SQL. Is this acceptable as long as the exception type is unchanged?

Reply:
I feel the more important problem here is what Shixuan worries: `generate_source_code` might not be reliable on some edge cases so if we raise an new exception because `generate_source_code` fails to capture the correct source code, then it will become very confusing?

Reply:
Jianzhun and I had a discussion on this, we feel we could be conservative for now by just using `inspect.getsource` to extract the source code of the function itself instead of using `generate_source_code`.

though in this case we lack the support for functions using global vars, other modules, but it's still some progress than nothing and once the `generate_source_code` is stable, we could shift to it.

as for the error type,  since we're already raising error which is a breaking change compared to the previous behavior which doesn't raise at all, is it a good opportunity to start using `SnowparkClientException`?

Reply:
Is it possible to make the exception type unchanged? If the current behavior is a UDF registration failure, then I expect the exception is more like a `ProgrammingError`? I'm not sure making the exception a `ProgrammingError` makes sense here.

If it has to be a behavior change, then I agree that it is a good idea to use `SnowparkClientException`. Also, I agree using `inspect.getsource` is a good compromise to avoid using experimental feature by default :)

Reply:
@sfc-gh-jdu Agree that we shouldn't use unstable code for production.

Reply:
Even we make exception type unchanged, the exception now will happen at different time: previously we get the exception during udf invocation, but now we get the exception during udf creation, which is very similar to https://github.com/snowflakedb/snowpark-python/commit/8f58b586e8a7e2b65ee6c87e44185ad25dea8b59. We regarded it as the behavior change before.